pkg/graveler/settings/manager.go

```diff
@@ -0,0 +1,125 @@
+	committedBlockStoragePrefix string
+}
+
+// Save persists the given proto.Message under the given repository and key.
+func (m *Manager) Save(ctx context.Context, repositoryID graveler.RepositoryID, key string, message proto.Message) error {
+	repo, err := m.refManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return err
+	}
+	messageBytes, err := proto.Marshal(message)
+	if err != nil {
+		return err
+	}
+	err = m.blockAdapter.Put(ctx, block.ObjectPointer{
+		StorageNamespace: string(repo.StorageNamespace),
+		Identifier:       fmt.Sprintf(settingsSuffixTemplate, m.committedBlockStoragePrefix, key),
+		IdentifierType:   block.IdentifierTypeRelative,
+	}, int64(len(messageBytes)), bytes.NewReader(messageBytes), block.PutOpts{})
+	if err != nil {
+		return err
+	}
+	logger := logging.FromContext(ctx)
+	if logger.IsTracing() {
+		logger.
+			WithField("repo", repositoryID).
+			WithField("key", key).
+			WithField("setting", protojson.Format(message)).
+			Trace("saving repository-level setting")
+	}
+	return nil
+}
+
+// Get fetches the configuration under the given repository and key.
+// The result is placed in message.
+func (m *Manager) Get(ctx context.Context, repositoryID graveler.RepositoryID, key string, message proto.Message) error {
+	repo, err := m.refManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return err
+	}
+	objectPointer := block.ObjectPointer{
+		StorageNamespace: string(repo.StorageNamespace),
+		Identifier:       fmt.Sprintf(settingsSuffixTemplate, m.committedBlockStoragePrefix, key),
+		IdentifierType:   block.IdentifierTypeRelative,
+	}
+	reader, err := m.blockAdapter.Get(ctx, objectPointer, -1)
+	if errors.Is(err, adapter.ErrDataNotFound) {
+		return graveler.ErrNotFound
+	}
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = reader.Close()
+	}()
+	messageBytes, err := io.ReadAll(reader)
+	if err != nil {
+		return err
+	}
+	err = proto.Unmarshal(messageBytes, message)
+	if err != nil {
+		return err
+	}
+	logger := logging.FromContext(ctx)
+	if logger.IsTracing() {
+		logger.
+			WithField("repo", repositoryID).
+			WithField("key", key).
+			WithField("setting", protojson.Format(message)).
+			Trace("got repository-level setting")
+	}
+	return nil
+}
+
+// UpdateWithLock atomically performs the following sequence:
+// 1. Fetch the configuration under the given repository and key, and place it in message.
+// 2. Call the given update function (which should presumably update the message).
+// 3. Persist message to the store.
+func (m *Manager) UpdateWithLock(ctx context.Context, repositoryID graveler.RepositoryID, key string, message proto.Message, update func()) error {
+	repo, err := m.refManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return err
+	}
+	_, err = m.branchLock.MetadataUpdater(ctx, repositoryID, repo.DefaultBranchID, func() (interface{}, error) {
+		err = m.Get(ctx, repositoryID, key, message)
+		if err != nil {
+			return nil, err
+		}
+		update()
```

Comment:
Shouldn't update get the read value?

Reply:
The value is loaded into the given `message` variable, and `update` should use it.

Reply:
I also do not understand how `update` can affect the message.  We get `message` from m, then we call `update` _which cannot see `message`_, then we save the exact `message` that we just loaded.